### DISCUSSION TITLE
Drawing a specific part of an image

### DISCUSSION THREAD

**Discussion by markand** [RudolphRiedel/FT800-FT813#122] [General]
Hey,

I don't know if that seems complicated or if I miss something obvious. When I use SDL on desktop apps I create an image with a predefined set of tiles and I crop a specific square that I draw on the screen so that you can load only one large image rather than individual bitmaps.

It looks like we can use either translate or scissors function but I'm unable to draw my clipped rectangle on the desired position.

For the example, using this 80x80 image:

![grid](https://github.com/RudolphRiedel/FT800-FT813/assets/11919008/85e81fd1-36e1-4078-823a-841164cfa71b)

I would like to cut the purple part and draw it at position 0x0 on the screen. for that, I can either use translate or scissor but in both case I get the same result. It's only the purple part that is shown on the screen but exactly like if the whole image was draw at 0x0. Ending with this output on my display:

![IMG_2799](https://github.com/RudolphRiedel/FT800-FT813/assets/11919008/ebbc2bd8-371e-4c49-842b-30bd1ff00ee3)

```
// trimmed down, but basically like this.
EVE_cmd_dl(SCISSOR_XY(40, 0));
EVE_cmd_dl(SCISSOR_SIZE(40, 40));
```

Any idea or documentations example I can follow? On the BRT_AN_033 note I have not found/understood how to do to that.

**Comment by RudolphRiedel** [comment]
I never really used SCISSOR_XY or SCISSOR_SIZE as usually EVE works more the opposite way, use tiles to compose a bigger picture.

The image is still drawn at 0/0, SCISSORS is just a window to that canvas.
You need to change the base coordinates for VERTEX2F, in this case you need to go negative for X.
So first place the image so that the tile is where you want it to be, then cut free the portion of the canvas that you want to see.

And you likely want to use SAVE_CONTEXT / RESTORE_CONTEXT as well.

**Reply by markand** [reply]
Thanks I was thinking of doing that even though I thought it was strange at a glance. In the programming manual I also see that we can use setmatrix and friends but this library has getmatrix only?

but those matrix computations are a bit complicated to me in contrast to the scissor feature. on the other hand the translate functions offer more features like scaling and rotating obviously.

**Reply by RudolphRiedel** [reply]
CMD_SETMATRIX and CMD_GETMATRIX are a bit odd.
Together with BITMAP_TRANSFORM_A to BITMAP_TRANSFORM_F.

I am still not really sure that I implement these correctly.
But well, CMD_SETMATRIX does not have parameters.
And the paramters of CMD_GETMATRIX are document as "output parameter".
Although the text right above this reads "This command retrieves the current matrix..."

Plus yes, BITMAP_TRANSFORM_A(val) needs a uint32_t but the actual parameter is defined as a "signed 15.8-bit fixed-point" and the field has 24 bits - whatever this is supposed to actually mean since the programming guide does not tell.

Next we have the "bitmap transformation matrix" and the "current matrix" in the context of the coprocessor engine.

I kind of gave up on these, hoping for more documentation some day, never really missed these.

**Reply by L-Storm** [reply]
Hi Rudolph, Love your Library! I've come across this comment when experiencing some issues using the BITMAP_TRANSFORM_A-F cmds.  I am trying to shrink and inflate bitmaps that I am loading from flash. When using the BITMAP_TRANSFORM_A-F I am struggling to get a display that looks correct. 

I am just wondering, if you gave up on these functions is there a different approach to achieve shrinking and inflating of images from flash? 


for reference this works as you would hope loading an image that is 1:1 scale: 

```
    uint32_t scale;
    scale = 1;
    cmd(SCISSOR_XY(x,y));
    cmd(SCISSOR_SIZE(iw*scale,ih*scale));
    cmd(CLEAR(1,1,1));

    cmd_setbitmap((address / 32) | 0x800000, RGBA_ASTC_4x4, iw, ih );
    cmd(BITMAP_TRANSFORM_E(256/scale)); //vertical axis stretch
    cmd(BITMAP_TRANSFORM_A(256/scale)); //horisontal axis stretch

    cmd(BEGIN(BITMAPS));
    cmd(VERTEX2II(x,y,0,0));
    
```
 But if you increase the scale factor, you get pretty strange results. I am using an EVE BT817 chip.

**Reply by RudolphRiedel** [reply]
Hi,

I am usually not using the BITMAP_TRANSFORM values directly.
For rotation there is CMD_ROTATE and for scaling there is CMD_SCALE.

In the examples CMD_SCALE is used as a zoom, but I am fairly certain that it should be possible to display an image at twice the size for example, propably with an extra BITMAP_SIZE command - have not tried to scale an image for quite some time though.
